### PR TITLE
Make Game::ObjectFadeAnimation synchronous to perform events in the correct order

### LINKED ISSUE #1151: Hero enters object before creature disappears
When hero attacks neutral creature in a cell, some object placed at, he firstly enters object, after battle, and only after he exit - creature disappears.
Attacking creature 3 0

### LINKED ISSUE #3074: Adventure map, defeated creature should be vanished before hero disappears in "Stone Liths"
Hero attacks peasants.
image
After victory hero changes his location. But peasants vanish only after hero appears in the exit "Stone Liths"
image
image

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
fix #3074
fix #1151

Needs to be thoroughly tested on different types of disappearing objects (neutral armies, resources, artifacts, chests, jails, barriers, flotsams, shipwreck survivors, and so on, as well as boats in case of "Summon Boat" spell).

### CODE REVIEW

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , is it ready for review?

**Comment by oleg-derevenetz** [comment]
Hi @ihhub Yes, I think so. There were just minor style changes recently. But I don't think that this PR should be merged in 0.9.2, it should be tested first.

File: src/fheroes2/heroes/heroes_action.cpp [hunk 1]
```diff
@@ -1819,10 +1839,14 @@ void ActionToArtifact( Heroes & hero, int obj, s32 dst_index )
 
         if ( result && hero.PickupArtifact( art ) ) {
             Game::PlayPickupSound();
-            AnimationRemoveObject( tile );
+
+            const uint8_t tileObj = tile.GetObject();
+            const uint32_t tileIdx = tile.GetIndex();
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes_action.cpp:1844]: `GetIndex` returns **int32_t**. Here we are introducing compilation warning.

File: src/fheroes2/heroes/heroes_action.cpp [hunk 2]
```diff
@@ -1900,11 +1924,15 @@ void ActionToTreasureChest( Heroes & hero, u32 obj, s32 dst_index )
         hero.GetKingdom().AddFundsResource( Funds( Resource::GOLD, gold ) );
 
     Game::PlayPickupSound();
-    AnimationRemoveObject( tile );
+
+    const uint8_t tileObj = tile.GetObject();
+    const uint32_t tileIdx = tile.GetIndex();
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes_action.cpp:1929]: `GetIndex` returns **int32_t**. Here we are introducing compilation warning.

File: src/fheroes2/heroes/heroes_action.cpp [hunk 3]
```diff
@@ -2925,10 +2953,14 @@ void ActionToJail( const Heroes & hero, u32 obj, s32 dst_index )
             _( "In a dazzling display of daring, you break into the local jail and free the hero imprisoned there, who, in return, pledges loyalty to your cause." ),
             Font::BIG, Dialog::OK );
 
-        AnimationRemoveObject( tile );
+        const uint8_t tileObj = tile.GetObject();
+        const uint32_t tileIdx = tile.GetIndex();
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes_action.cpp:2957]: `GetIndex` returns **int32_t**. Here we are introducing compilation warning.

File: src/fheroes2/heroes/heroes_action.cpp [hunk 4]
```diff
@@ -3058,10 +3090,15 @@ void ActionToBarrier( Heroes & hero, u32 obj, s32 dst_index )
             _( "A magical barrier stands tall before you, blocking your way. Runes on the arch read,\n\"Speak the key and you may pass.\"\nAs you speak the magic word, the glowing barrier dissolves into nothingness." ),
             Font::BIG, Dialog::OK );
 
-        AnimationRemoveObject( tile );
+        const uint8_t tileObj = tile.GetObject();
+        const uint32_t tileIdx = tile.GetIndex();
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes_action.cpp:3094]: `GetIndex` returns **int32_t**. Here we are introducing compilation warning.

File: src/fheroes2/game/game.cpp [hunk 5]
```diff
@@ -231,51 +231,107 @@ void Game::SetCurrentMusic( int mus )
     current_music = mus;
 }
 
-void Game::ObjectFadeAnimation::FinishFadeTask()
+void Game::ObjectFadeAnimation::PerformFadeTask( uint8_t object, uint32_t fromIndex, uint32_t toIndex, bool fadeOut, bool fadeIn )
```
Comment (ihhub) [inline src/fheroes2/game/game.cpp:234]: Tile indexes are `int32_t` type. Could you please change them?

File: src/fheroes2/heroes/heroes_action.cpp (hunk 1 above)
Comment (ihhub) [reply src/fheroes2/heroes/heroes_action.cpp:1844]: and actually in the rest of places as well.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @oleg-derevenetz , I left multiple comments in relation to type conversion. Also I quickly tested the version and I found that after a battle we display an incorrect monster frame disappering on the map. On the screenshot I attacked Pikemen:
image

**Comment by oleg-derevenetz** [comment]
Hi @ihhub

> Tile indexes are int32_t type. Could you please change them?

That's probably because Tiles::maps_index is uint32_t. So if we change signedness here, we will get a warning in another place.

**Comment by ihhub** [comment]
@oleg-derevenetz , thank you so much for fixing these issues!

**Pull request merged**